### ISSUE TITLE
MonthDay range doesn't properly adjust ending_year/current_year 

### ISSUE THREAD

**Issue by boc-tothefuture** [boc-tothefuture/openhab-jruby#479]
This code:
```
# Wrap to next year if ending day of month is before starting day of month
ending_year = ending < start ? next_year : current_year
```
Doesn't work correctly. 
1. It adjust ending year when it shouldn't -- right now it always increments the ending year even in cases when it shouldn't. e.g. 12/31 and 1/2 when testing on today (1/1) should be 12/31/2021 and 1/2/2022.  Instead it is 12/31/2022 and 1/2/2023.
2. There is no method to adjust backwards the starting_year. 

The current year needs to play into the decision about the start and end year. 

I don't quite know what the algorithm should be.  @ccutrer or @jimtng not sure if either of you have ideas on what the algorithm is.  I looked at it for a few minute and it didn't jump off the page to me.  I might come back after a break and see if that helps.

**Comment by jimtng** [comment]
Like in TimeOfDay, MonthDayRangeElement needs to know range_begin, so MonthDayRangeElement#<=> can add a year to the "other" if it's less than range_begin

**Comment by jimtng** [comment]
I'd also like to see MonthDay#between? implemented like TimeOfDay

**Comment by jimtng** [comment]
Possible cucumber test below. Not sure if we need to also test TimeOfDay, perhaps it's already done elsewhere, and the commented tests passed.

```
  Scenario Outline: Between supports range rollover
    Given a rule template:
      """
      expr = "<type>.parse('<value>').between?(<range>)"
      result = eval(expr)
      logger.info("between? #{result}")
      """
    When I deploy the rule
    Then It should log "between? <result>" within 5 seconds
    Examples: Checks range
      | type     | value | range            | result |
      # | TimeOfDay | 23:00 | '20:00'..'02:00' | true   |
      # | TimeOfDay | 19:00 | '20:00'..'02:00' | false  |
      # | TimeOfDay | 1:00  | '20:00'..'02:00' | true   |
      # | TimeOfDay | 3:00  | '20:00'..'02:00' | false  |
      | MonthDay | 12-25 | '12-01'..'01-05' | true   |
      | MonthDay | 11-25 | '12-01'..'01-05' | false  |
      | MonthDay | 01-01 | '12-01'..'01-05' | true   |
      | MonthDay | 01-20 | '12-01'..'01-05' | false  |

```

**Comment by boc-tothefuture** [comment]
> I'd also like to see MonthDay#between? implemented like TimeOfDay

Not sure I follow? It is a range and the range class implements `between?`

**Comment by jimtng** [comment]
MonthDay.now.between?('05-02'..'05-10')

**Comment by jimtng** [comment]
Instead of the test above, WDYT of the following tests, perhaps located inside month_date.feature rather than between.feature?

Note that currently PR #483 failed on tests `MonthDay range handles invalid range`

```
  Scenario Outline: MonthDay supports range
    Given a rule template:
      """
      result = <type>.parse('<value>').between?(<range>)
      logger.info("between? #{result}")
      """
    When I deploy the rule
    Then It should log "between? <result>" within 5 seconds
    Examples: Checks range
      | type     | value | range             | result |
      # normal range
      | MonthDay | 01-20 | '12-01'..'12-05'  | false  |
      | MonthDay | 12-01 | '12-01'..'12-05'  | true   |
      | MonthDay | 12-03 | '12-01'..'12-05'  | true   |
      | MonthDay | 12-05 | '12-01'..'12-05'  | true   |
      | MonthDay | 12-05 | '12-01'...'12-05' | false  |
      | MonthDay | 12-06 | '12-01'..'12-05'  | false  |
      # same day range
      | MonthDay | 12-01 | '12-02'..'12-02'  | false  |
      | MonthDay | 12-02 | '12-02'..'12-02'  | true   |
      | MonthDay | 12-03 | '12-02'..'12-02'  | false  |

  Scenario Outline: MonthDay range supports Feb 29th
    Given a rule template:
      """
      result = <type>.parse('<value>').between?(<range>)
      logger.info("between? #{result}")
      """
    When I deploy the rule
    Then It should log "between? <result>" within 5 seconds
    Examples: Checks range
      | type     | value | range             | result |
      # leap year
      | MonthDay | 02-29 | '02-01'..'03-01'  | true   |
      | MonthDay | 02-29 | '02-01'...'03-01' | true   |

  Scenario Outline: MonthDay range can span end of year
    Given a rule template:
      """
      result = <type>.parse('<value>').between?(<range>)
      logger.info("between? #{result}")
      """
    When I deploy the rule
    Then It should log "between? <result>" within 5 seconds
    Examples: Checks range
      | type     | value | range             | result |
      # crossing over the end of the year
      | MonthDay | 11-25 | '12-01'..'01-05'  | false  |
      | MonthDay | 12-01 | '12-01'..'01-05'  | true   |
      | MonthDay | 12-25 | '12-01'..'01-05'  | true   |
      | MonthDay | 01-01 | '12-01'..'01-05'  | true   |
      | MonthDay | 01-05 | '12-01'..'01-05'  | true   |
      | MonthDay | 01-05 | '12-01'...'01-05' | false  |
      | MonthDay | 01-20 | '12-01'..'01-05'  | false  |

  Scenario Outline: MonthDay range handles invalid range
    Given a rule template:
      """
      result = <type>.parse('<value>').between?(<range>)
      logger.info("between? #{result}")
      """
    When I deploy the rule
    Then It should log "between? <result>" within 5 seconds
    Examples: Checks range
      | type     | value | range             | result |
      # invalid range
      | MonthDay | 11-01 | '12-02'..'12-01'  | false  |
      | MonthDay | 12-01 | '12-02'..'12-01'  | false  |
      | MonthDay | 12-02 | '12-02'..'12-01'  | false  |
      | MonthDay | 12-02 | '12-02'...'12-02' | false  |
      | MonthDay | 12-03 | '12-02'..'12-02'  | false  |

```

**Comment by jimtng** [comment]
the "invalid" range, is actually a case of range roll over. Do you think it should be treated as such?

**Comment by boc-tothefuture** [comment]
I can take those tests and move them into monthday.

The PR currently seems to be passing?

**Comment by boc-tothefuture** [comment]
@jimtng what is invalid about those ranges? '12-02' to '12-01' covers every day of the year after it wraps around.  All of those ranges are valid.

**Comment by jimtng** [comment]
> @jimtng what is invalid about those ranges? '12-02' to '12-01' covers every day of the year after it wraps around.  All of those ranges are valid. 
> 

You're correct, they're valid. I was looking at normal integer ranges and (5..1).covers? Returns false for everything and I didn't realise that for MonthDay it's a case of a wraparound/rollover range.
